### PR TITLE
Add more Notebook telemetry events

### PR DESCRIPTION
**Pull request by MaddyDev**
More notebook telemetry capturing the below actions:

1. Search Book through Search viewlet -> launching search from treeItem actions.
2. Move Book
3. Run All Cells
4. Add Cell
5. Change Kernel
6. Search in Notebooks Viewlet
7. Search Completed with time taken
8. New Notebook launched from Connections.

### CODE REVIEW

**Comment by chlafreniere** [comment]
cc @udeeshagautam @qifahs FYI

File: src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookExplorerViewlet.ts [hunk 1]
```diff
@@ -314,6 +320,10 @@ export class NotebookExplorerViewPaneContainer extends ViewPaneContainer {
 				this.searchWidget.focus(false, true); // focus back to input field
 			}
 		}, onQueryValidationError);
+		let end = new Date().getTime();
+		this._telemetryService?.createActionEvent(TelemetryKeys.TelemetryView.Notebook, TelemetryKeys.NbTelemetryAction.SearchComplete)
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookExplorerViewlet.ts:324]: 1. This isn't fired when the search is complete - this function just kicks off the search and returns immediately. The actual search will complete at some point later
2. Typically we use camelCase for property names (`timeTaken`)
3. Time taken should be a measurement - not a property
4. All measurements should have their units in the name (such as `timeTakenMs`)

Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookExplorerViewlet.ts:324]: You don't want to capture any additional information such as # of results returned, how many different files those results were in, etc?

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: src/sql/platform/telemetry/common/telemetryKeys.ts [hunk 2]
```diff
@@ -79,6 +79,11 @@ export enum TelemetryAction {
 
 export enum NbTelemetryAction {
 	RunCell = 'RunCell',
-	RunAll = 'RunNotebook'
+	RunAll = 'RunNotebook',
+	AddCell = 'AddCell',
+	ChangeKernel = 'ChangeKernel',
+	NewNotebookFromConnections = 'NewNotebookWithConnectionProfile',
+	SearchStarted = 'SearchStarted',
```
Comment (Charles-Gagnon) [inline src/sql/platform/telemetry/common/telemetryKeys.ts:86]: These seem a bit too generic to warrant being in `NbTelemetryAction` - would suggest putting in `TelemetryAction` enum and then just ensure that you have the view set properly in the events to know where the search was triggered from.

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 3]
```diff
@@ -213,12 +219,14 @@ export class RunAllCellsAction extends Action {
...
 		@INotificationService private notificationService: INotificationService,
-		@INotebookService private _notebookService: INotebookService
+		@INotebookService private _notebookService: INotebookService,
+		@IAdsTelemetryService private _telemetryService?: IAdsTelemetryService,
 	) {
 		super(id, label, cssClass);
 	}
 	public async run(context: URI): Promise<boolean> {
 		try {
+			this._telemetryService?.sendActionEvent(TelemetryKeys.TelemetryView.Notebook, TelemetryKeys.NbTelemetryAction.RunAll);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookActions.ts:229]: You can remove the ? from all these now

File: src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookSearch.ts [hunk 4]
```diff
@@ -253,7 +257,11 @@ export class NotebookSearchView extends SearchView {
 		}, 2000);
 
 		const onComplete = async (completed?: ISearchComplete) => {
-			clearTimeout(slowTimer);
+			let end = new Date().getTime();
+			this._telemetryService?.createActionEvent(TelemetryKeys.TelemetryView.Notebook, TelemetryKeys.NbTelemetryAction.SearchCompleted)
+				.withAdditionalProperties({ timeTakenMs: end - start, resultsReturned: completed.results.length })
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookSearch.ts:262]: timeTakenMs should still be a measurement - `withAdditionalMeasurements`

File: src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookSearch.ts [hunk 5]
```diff
@@ -253,7 +257,11 @@ export class NotebookSearchView extends SearchView {
...
+				.send(); clearTimeout(slowTimer);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookSearch.ts:263]: clearTimeout should be on new line

File: src/sql/workbench/services/notebook/browser/models/cell.ts [hunk 6]
```diff
@@ -88,10 +88,10 @@ export class CellModel extends Disposable implements ICellModel {
 
 	constructor(cellData: nb.ICellContents,
 		private _options: ICellModelOptions,
+		@IAdsTelemetryService private _telemetryService?: IAdsTelemetryService,
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/cell.ts:91]: Just undo this change - the services shouldn't be optional here but no reason to change this now that I can see

File: src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookExplorerViewlet.ts [hunk 7]
```diff
@@ -255,6 +258,9 @@ export class NotebookExplorerViewPaneContainer extends ViewPaneContainer {
 			onQueryValidationError(err);
 			return;
 		}
+		this._telemetryService?.createActionEvent(TelemetryKeys.TelemetryView.Notebook, TelemetryKeys.NbTelemetryAction.SearchStarted)
+			.withAdditionalProperties({ triggered_on_type: triggeredOnType })
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookExplorerViewlet.ts:262]: camelCase for this - `triggeredOnType`

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 8]
```diff
@@ -549,13 +562,16 @@ export class NewNotebookAction extends Action {
...
-		@IObjectExplorerService private objectExplorerService: IObjectExplorerService
+		@IObjectExplorerService private objectExplorerService: IObjectExplorerService,
+		@optional(IAdsTelemetryService) private _telemetryService?: IAdsTelemetryService,
 	) {
 		super(id, label);
 		this.class = 'notebook-action new-notebook';
 	}
 
 	async run(context?: azdata.ObjectExplorerContext): Promise<void> {
+		this._telemetryService?.createActionEvent(TelemetryKeys.TelemetryView.Notebook, TelemetryKeys.NbTelemetryAction.NewNotebookFromConnections)
```
[the remark this answers is not in the document]
Comment (MaddyDev) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:572]: am trying to add if the connection is onPrem or Azure here. Do you know which property from the connectionProfile tells me that?

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 3 above)
Comment (MaddyDev) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:229]: What if the _telemetryService parameter is passed as undefined (as I did in some tests). I think it's foolproof to have that.

File: src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookExplorerViewlet.ts (hunk 7 above)
Comment (MaddyDev) [reply src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookExplorerViewlet.ts:262]: missed this one :(

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 3 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:229]: It should never be - for tests use https://github.com/microsoft/azuredatastudio/blob/main/src/sql/platform/telemetry/common/adsTelemetryService.ts#L217

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:572]: I've added 2 helper functions for getting additional connection information : 

https://github.com/microsoft/azuredatastudio/blob/main/src/sql/platform/telemetry/common/adsTelemetryService.ts#L47
https://github.com/microsoft/azuredatastudio/blob/main/src/sql/platform/telemetry/common/adsTelemetryService.ts#L56

File: src/sql/workbench/contrib/notebook/browser/notebookActions.ts [hunk 9]
```diff
@@ -346,6 +356,9 @@ export class KernelsDropdown extends SelectBox {
 	}
 
 	public doChangeKernel(displayName: string): void {
+		this._telemetryService?.createActionEvent(TelemetryKeys.TelemetryView.Notebook, TelemetryKeys.NbTelemetryAction.ChangeKernel)
+			.withAdditionalProperties({ kernel: displayName })
```
[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:359]: @MaddyDev This is still using display name - was that done on purpose?

[the remark this answers is not in the document]
Comment (MaddyDev) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:359]: Yeah, am not sure what you meant by shouldn't use displayName, that's the only variable available in the scope.

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/notebookActions.ts:359]: Didn't look closely enough at the code. 

The reasoning was stated - we shouldn't sending values that may be localized.

I would suggest moving this down to https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/services/notebook/browser/models/notebookModel.ts#L931 or maybe even lower down. Although you may want to consider adding a source parameter to the event then to tell you what was changing the kernel if you care about differentiating that.

File: src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookExplorerViewlet.ts [hunk 10]
```diff
@@ -124,7 +126,8 @@ export class NotebookExplorerViewPaneContainer extends ViewPaneContainer {
 		@IMenuService private menuService: IMenuService,
 		@IContextKeyService private contextKeyService: IContextKeyService,
 		@IViewDescriptorService viewDescriptorService: IViewDescriptorService,
-		@IFileService private readonly fileService: IFileService
+		@IFileService private readonly fileService: IFileService,
+		@IAdsTelemetryService private _telemetryService?: IAdsTelemetryService
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookExplorerViewlet.ts:130]: This shouldn't be optional

File: src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookSearch.ts [hunk 11]
```diff
@@ -82,6 +84,7 @@ export class NotebookSearchView extends SearchView {
 		@IOpenerService openerService: IOpenerService,
 		@ITelemetryService telemetryService: ITelemetryService,
 		@ICommandService readonly commandService: ICommandService,
+		@IAdsTelemetryService private _telemetryService?: IAdsTelemetryService,
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookSearch.ts:87]: This shouldn't be optional

File: src/sql/workbench/services/notebook/browser/models/notebookModel.ts [hunk 12]
```diff
@@ -888,6 +888,8 @@ export class NotebookModel extends Disposable implements INotebookModel {
 	public changeKernel(displayName: string): void {
 		this._selectedKernelDisplayName = displayName;
 		this._currentKernelAlias = this.context?.serverCapabilities?.notebookKernelAlias;
+		this.adstelemetryService.createActionEvent(TelemetryKeys.TelemetryView.Notebook, TelemetryKeys.NbTelemetryAction.ChangeKernel)
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/notebookModel.ts:891]: You're still using the localized displayName here. Is there something preventing you from putting this somewhere like line 924 after we get the IKernelSpec for the selected kernel? (or even in the client session instead)

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

**Review by Charles-Gagnon — APPROVED** [review]
I'll merge this in and get the port request opened once it passes the checks

**Pull request merged**